### DISCUSSION TITLE
dynamic state feature is not supported on the Android platform.

### DISCUSSION THREAD

**Discussion by dxd39** [MethanePowered/MethaneKit#126] [Q&A] [answered]
hello:
I encountered issues while trying to implement android platform in [my fork](https://github.com/dxd39/MethaneKit/commit/9b2a0b6ae44f456a3d01bb3fbab9504f8ea1737e) . There are some extensions that are not supported on the Android.  I used  [Vulkan-ExtensionLayer](https://github.com/KhronosGroup/Vulkan-ExtensionLayer) fix some of them(synchronization2 and timeline_semaphore). However, dynamic state still not resolved. Do you have a plan to fix it or any compatibility solutions? 
<img width="1920" alt="android" src="https://github.com/MethanePowered/MethaneKit/assets/10190265/3c5b3667-f579-4df4-b06b-5a1375a64b0f">

**Comment by egorodet** [comment]
Hello @dxd39,

First of all let me say Thank You for your work on Android support in this project! I really appreciate your contribution efforts!

Methane Kit is designed to extensively use `RenderCommandList::SetViewState` for text rendering and some UI rendering routines. Also viewports and scissor rects setup is split from PSO state in DirectX 12 and Metal, so I wanted to have a similar solution for Vulkan. In Vulkan similar API behavior is possible only via dynamic state extension, that's why it was used. Unfortunately I did not know that this extension is not supported on Android... I also discovered that this extension is not available for MoltenVK on MacOS.

At the moment, I don't have a recommendation how to fix it. Let me think about possible solutions and come back to you later.

**Comment by dxd39** [comment]
I have done some additional testing, to be precise, the Android platform does not support dynamic state extension or new features after version 1.1, but supports some basic dynamic state features, so some compatibility processing can be done.
<img width="1262" src="https://github.com/MethanePowered/MethaneKit/assets/10190265/b0730b4d-f5a6-451a-a1d2-c44bc4b8a97f">

**Reply by egorodet** [reply]
Yes, you're right. I can try to add compatibility support of monolithic PSO for `Vulkan::RenderState` in case when dynamic state extension is not available, but with a cost of some performance overhead.

**Reply by egorodet** [reply]
I'm currently working on monolythic Pipeline state support for Vulkan. It will be ready in a couple of days.

**Comment by egorodet** [comment]
@dxd39 Initial support of the monolithic pipeline state for Vulkan is available in this branch `hotfix/vulkan-monolithic-state-support`. It is not fully tested yet, but it allows to run Vulkan "HelloTriangle" tutorial on MacOS via MoltenVK, which was not possible before. There is some problem with resources upload in other more complex tutorials when started via MoltenVK, which also needs to be resolved. I will let you know when this work is finished and merged in `master` branch.

**Reply by dxd39** [reply]
@egorodet Great! I can use this fix to continue testing on Android.

**Comment by egorodet** [comment, marked as answer]
@dxd39 Vulkan monolithic pipeline state support is now ready, validated and merged to `master` branch along with some fixes of the Vulkan validation layer errors (see MR #127 for more details). I found some occasional crashes, which could be signs of memory corruption on Linux platform, but due to Valgrind crashing itself I was not able to fix it yet. I hope that it will not block your Android porting efforts.

**Reply by dxd39** [reply]
Great, I will keep trying and see how it goes.
